### PR TITLE
feat: Add Non Admin Restore controller

### PR DESCRIPTION
**Pull request by mateusoliveira43**
## Why the changes were made

Non admin users want to restore NonAdminBackups. This PR adds NonAdminRestore CRD along with its controller.

Related to https://github.com/migtools/oadp-non-admin/issues/35

~Blocked by https://github.com/migtools/oadp-non-admin/pull/16~
~Blocked by https://github.com/migtools/oadp-non-admin/pull/54~
Blocked by https://github.com/migtools/oadp-non-admin/pull/67
Blocked by https://github.com/openshift/oadp-operator/pull/1388

## How the changes were made

Get kubebuilder version `3.14.0` from https://github.com/kubernetes-sigs/kubebuilder/releases and run command documented in Architecture https://github.com/migtools/oadp-non-admin/commit/62d521341c6c10310eecc96be9291327e66ec9b2

Run `make manifests`

Fix code quality issues

## How to test the changes made

TODO

## TODO

This PR still misses
- write restore logic
- [check PR size, can be separate] Adding example of configuring DPA with restricted NonAdminRestore fields https://github.com/migtools/oadp-non-admin/blob/94c4be9bf93da2d791281d2aaca54655d2821316/README.md?plain=1#L29
- https://github.com/migtools/oadp-non-admin/blob/94c4be9bf93da2d791281d2aaca54655d2821316/README.md?plain=1#L50
  - show user how to see NonAdminRestore status
  - show user how to see NonAdminRestore queue 
- unit/integration tests for new controller. Some ideas
  - [ERROR] try to create NonAdminRestore from Velero Backup
  - [ERROR] try to create NonAdminRestore from non existing NonAdminBackup
  - [ERROR] try to create NonAdminRestore NonAdminBackup with restricted fields
  - [ERROR] try to create NonAdminRestore in non existing namespace
  - [SUCCESS] create NonAdminRestore from NonAdminBackup owned by user

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -48,6 +48,7 @@ help: ## Display this help.
 .PHONY: manifests
 manifests: controller-gen ## Generate WebhookConfiguration, ClusterRole and CustomResourceDefinition objects.
 	$(CONTROLLER_GEN) rbac:roleName=non-admin-controller-role crd webhook paths="./..." output:crd:artifacts:config=config/crd/bases
+	sed -i 's/Velero backup/NonAdminBackup/' ./config/crd/bases/nac.oadp.openshift.io_nonadminrestores.yaml
```
Comment (mateusoliveira43) [inline Makefile:51]: for better user experience

File: cmd/main.go [hunk 2]
```diff
@@ -98,7 +98,8 @@ func main() {
 		TLSOpts: tlsOpts,
 	})
 
-	if len(constant.OadpNamespace) == 0 {
+	// TODO create get function in common :question:
```
Comment (mateusoliveira43) [inline cmd/main.go:101]: TODO

File: internal/controller/suite_test.go [hunk 3]
```diff
@@ -73,6 +73,9 @@ var _ = ginkgov2.BeforeSuite(func() {
 	err = nacv1alpha1.AddToScheme(scheme.Scheme)
 	gomega.Expect(err).NotTo(gomega.HaveOccurred())
 
+	err = nacv1alpha1.AddToScheme(scheme.Scheme)
```
Comment (mateusoliveira43) [inline internal/controller/suite_test.go:76]: seems duplication, check if it is needed

File: README.md [hunk 4]
```diff
@@ -47,7 +49,28 @@ To use NAC functionality:
...
+    - create NonAdminRestore
+
+        For example, use one of the sample NonAdminRestore available in `hack/samples/restores/` folder, by running
+        ```sh
+        oc process -f ./hack/samples/restores/<type> \
+            -p NAMESPACE=<non-admin-user-namespace> \
+            -p NAME=<NonAdminBackup-name> \
+            | oc create -f -
+        ```
+        <!-- TODO how to track status -->
```
Comment (mateusoliveira43) [inline README.md:71]: TODO

File: api/v1alpha1/nonadminrestore_types.go [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+	LogLevel string `json:"logLevel,omitempty"`
+	// TODO ALSO ADD TEST FOR DIFFERENT LOG LEVELS
+}
+
+// NonAdminRestoreStatus defines the observed state of NonAdminRestore
+type NonAdminRestoreStatus struct {
+	// Related Velero Restore name.
+	// +optional
+	VeleroRestoreName string `json:"veleroRestoreName,omitempty"`
+
```
Comment (mateusoliveira43) [inline api/v1alpha1/nonadminrestore_types.go:44]: do we need velero restore namespace :question: I think we already have this information here https://github.com/migtools/oadp-non-admin/blob/master/internal/common/constant/constant.go#L35-L41

File: config/crd/bases/nac.oadp.openshift.io_nonadminbackups.yaml [hunk 6]
```diff
@@ -589,8 +589,8 @@ spec:
                   type: object
                 type: array
               phase:
-                description: NonAdminBackupPhase is a simple one high-level summary
-                  of the lifecycle of an NonAdminBackup.
+                description: NonAdminPhase is a simple one high-level summary of the
```
Comment (mateusoliveira43) [inline config/crd/bases/nac.oadp.openshift.io_nonadminbackups.yaml:592]: TODO wrong description to user

File: internal/common/function/function.go [hunk 7]
```diff
@@ -127,7 +128,7 @@ func GenerateVeleroBackupName(namespace, nabName string) string {
 }
 
 // UpdateNonAdminPhase updates the phase of a NonAdminBackup object with the provided phase.
-func UpdateNonAdminPhase(ctx context.Context, r client.Client, logger logr.Logger, nab *nacv1alpha1.NonAdminBackup, phase nacv1alpha1.NonAdminBackupPhase) (bool, error) {
+func UpdateNonAdminPhase(ctx context.Context, r client.Client, logger logr.Logger, nab *nacv1alpha1.NonAdminBackup, phase nacv1alpha1.NonAdminPhase) (bool, error) {
```
Comment (mpryc) [inline internal/common/function/function.go:131]: This actually should be a separate change, so we don't mix restore with change of types. There is already common condition, but it's ok to move also phases to be similar. https://github.com/migtools/oadp-non-admin/blob/master/api/v1alpha1/nonadmincontroller_types.go#L21

File: internal/common/function/function.go [hunk 8]
```diff
@@ -159,7 +160,7 @@ func UpdateNonAdminPhase(ctx context.Context, r client.Client, logger logr.Logge
 // based on the provided parameters. It validates the input parameters and ensures
 // that the condition is set to the desired status only if it differs from the current status.
 // If the condition is already set to the desired status, no update is performed.
-func UpdateNonAdminBackupCondition(ctx context.Context, r client.Client, logger logr.Logger, nab *nacv1alpha1.NonAdminBackup, condition nacv1alpha1.NonAdminCondition, conditionStatus metav1.ConditionStatus, reason string, message string) (bool, error) {
+func UpdateNonAdminBackupCondition(ctx context.Context, r client.Client, logger logr.Logger, nab *nacv1alpha1.NonAdminBackup, condition types.NonAdminCondition, conditionStatus metav1.ConditionStatus, reason string, message string) (bool, error) {
```
Comment (mpryc) [inline internal/common/function/function.go:163]: We already have this as common:
  https://github.com/migtools/oadp-non-admin/blob/master/api/v1alpha1/nonadmincontroller_types.go#L21

File: internal/common/function/function.go [hunk 9]
```diff
@@ -29,12 +29,13 @@ import (
 	velerov1api "github.com/vmware-tanzu/velero/pkg/apis/velero/v1"
 	apimeta "k8s.io/apimachinery/pkg/api/meta"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
-	"k8s.io/apimachinery/pkg/types"
+	apitypes "k8s.io/apimachinery/pkg/types"
```
Comment (mpryc) [inline internal/common/function/function.go:32]: how about changing this to:
```
apimeta
apimetav1
apimetatypes
```

File: internal/common/types/types.go [hunk 10]
```diff
@@ -14,17 +14,8 @@ See the License for the specific language governing permissions and
 limitations under the License.
 */
 
-package v1alpha1
```
Comment (mpryc) [inline internal/common/types/types.go:17]: I don't know if that's valuable change. Really mixing the api with implementation types as some of them we have in api/* and now some will be in internal/* so a bit of mix that is harder to read. Definitely not part of the NAR controller, can be seaparate PR.

File: internal/controller/nonadminrestore_controller.go [hunk 11]
```diff
@@ -0,0 +1,120 @@
...
+	}
+
+	// TODO nonAdminRestore respect restricted fields
+
+	nonAdminBackupName := objectSpec.RestoreSpec.BackupName
+	nonAdminBackup := &nacv1alpha1.NonAdminBackup{}
+	err := r.Get(ctx, types.NamespacedName{Namespace: req.Namespace, Name: nonAdminBackupName}, nonAdminBackup)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			// TODO add this error message to NonAdminRestore status
```
Comment (mpryc) [inline internal/controller/nonadminrestore_controller.go:89]: There is similar implementation in PR for Backup, could be common func:

https://github.com/migtools/oadp-non-admin/pull/56/files#diff-adbb2c0cf906a4dc9b38604d85fb6f353a314513673ea181267a40625f80237bR50-R55

File: internal/controller/nonadminrestore_controller.go [hunk 12]
```diff
@@ -0,0 +1,120 @@
...
+// +kubebuilder:rbac:groups=nac.oadp.openshift.io,resources=nonadminrestores,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=nac.oadp.openshift.io,resources=nonadminrestores/status,verbs=get;update;patch
+// +kubebuilder:rbac:groups=nac.oadp.openshift.io,resources=nonadminrestores/finalizers,verbs=update
+
+// +kubebuilder:rbac:groups=velero.io,resources=restores,verbs=get;list;watch;create;update;patch
+
+// Reconcile is part of the main kubernetes reconciliation loop which aims to
+// move the current state of the NonAdminRestore to the desired state.
+func (r *NonAdminRestoreReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	r.Logger = log.FromContext(ctx)
```
Comment (mpryc) [inline internal/controller/nonadminrestore_controller.go:53]: This is bad.  We can assign one per worker but not every reconcile. Removed number of items from NAB reconciler, see:
https://github.com/migtools/oadp-non-admin/pull/56/files#diff-98b8a647a66c83f5ee56f93b7530fb37a28301390936ecdb1fdaf38c031453c6R38-R41

File: internal/common/types/types.go (hunk 10 above)
Comment (mateusoliveira43) [reply internal/common/types/types.go:17]: I see api folder as the entrypoint for kubebuilder to build CRD yamls, stored in config folder. So, there should be only things used to crate those CRDs in this folder

the type here, was not used by any CRD, so I do not think is the right place to it. Moved to internal common folder where we are storing all the things used more than once by files of the project

do you agree :question: 

I can do this in separate PR yes

File: api/v1alpha1/nonadminrestore_types.go [hunk 13]
```diff
@@ -0,0 +1,76 @@
...
+	// Related Velero Restore status.
+	// +optional
+	VeleroRestoreStatus *velerov1api.RestoreStatus `json:"veleroRestoreStatus,omitempty"`
+
+	Phase      NonAdminPhase      `json:"phase,omitempty"`
+	Conditions []metav1.Condition `json:"conditions,omitempty"`
+}
+
+// +kubebuilder:object:root=true
+// +kubebuilder:subresource:status
```
Comment (mateusoliveira43) [inline api/v1alpha1/nonadminrestore_types.go:54]: TODO add short name as in https://github.com/migtools/oadp-non-admin/pull/63

File: api/v1alpha1/nonadminrestore_types.go (hunk 5 above)
Comment (mateusoliveira43) [reply api/v1alpha1/nonadminrestore_types.go:42]: user may not have the info

**Comment by mateusoliveira43** [comment]
closing in favor of https://github.com/migtools/oadp-non-admin/pull/114

**Pull request closed without merging**